### DISCUSSION TITLE
How does record.Save() upserts?

### DISCUSSION THREAD

**Discussion by wsw70** [pocketbase/pocketbase#2182] [Q&A] [answered]
I have a limited knowledge of SQL (read: it's years I am considering learning it and started yesterday or so :)) and when reading the documentation for  [`func (*Dao) SaveRecord`](https://pkg.go.dev/github.com/pocketbase/pocketbase/daos#Dao.SaveRecord) I saw (emphasis mine)

> SaveRecord **upserts** the provided Record model

My limited understanding of "upsert" is that it either inserts or updates a record. I therefore thought that

```go
	record := models.NewRecord(collection)
	record.Set("id", userId)
        record.Set("mail", user.Mail)
	if err := app.Dao().SaveRecord(record); err != nil {
		log.Fatal().Msgf("cannot save %s in users: %v", user.Identificator, err)
	}
```

would update the field `mail` of the record with `id` equal to `userId` (which I know, and know that the record exists).

What I get is an error `UNIQUE constraint failed` which suggests that it is trying to **insert** the record (but fails because `id` is already there).

If I do not provide the `id`, a new record is created (which i not what I want to do - I want to update it).

What am I doing wrong?

**Comment by ganigeorgiev** [comment, marked as answer]
You can check https://github.com/pocketbase/pocketbase/discussions/2079.

`SaveRecord` is not the same as the SQL `UPSERT` statement. It performs `UPDATE` or `INSERT` based on the models `isNew()` return value.

We need to know upfront whether the record exist or not in order to trigger the correct before model hook - `OnModelBeforeCreate` or `OnModelBeforeUpdate` (this eventually may be simplified in the future).

**Reply by wsw70** [reply]
Thank you very much for the answer. I now understand this.

Would you mind if I did a PR for the documentation? (also about the attribute `name` which I found to be what you need to have so that it is displayed in the UI for a relationship?)

**Reply by ganigeorgiev** [reply]
> Would you mind if I did a PR for the documentation?

I'll consider updating the godoc sometime later today.

> also about the attribute name which I found to be what you need to have so that it is displayed in the UI for a relationship?

You can define an explicit `relation` display field from the `relation` field options in the Admin UI. If nothing is specified, we check for any commonly used display fields like name, title, email, username, label, key, etc.
